include/mbedtls/pkcs11_client.h

```diff
@@ -0,0 +1,150 @@
+
+#define MBEDTLS_PK_FLAG_SENSITIVE    ( (uint32_t) 0x00000001 )
+#define MBEDTLS_PK_FLAG_EXTRACTABLE  ( (uint32_t) 0x00000002 )
+#define MBEDTLS_PK_FLAG_SIGN         ( (uint32_t) 0x00000010 )
+#define MBEDTLS_PK_FLAG_VERIFY       ( (uint32_t) 0x00000020 )
+#define MBEDTLS_PK_FLAG_DECRYPT      ( (uint32_t) 0x00000040 )
+#define MBEDTLS_PK_FLAG_ENCRYPT      ( (uint32_t) 0x00000080 )
+
+#include "pk.h"
+
+/**
+ * \brief               Set up a PK context for a key pair in a PKCS#11 token
+ *
+ * \param ctx           PK context to fill, which must have been initialized
+ *                      with mbedtls_pk_init().
+ * \param hSession      Cryptoki session.
+ * \param hPublicKey    Cryptoki handle of the public key.
+ * \param hPrivateKey   Cryptoki handle of the private key, or
+ *                      CK_INVALID_HANDLE for a public key rather than a key
+ *                      pair.
+ *
+ * \return              0 on success,
+ *                      or MBEDTLS_ERR_PK_XXX error code.
+ *
+ * \note                The session and the key(s) must remain valid until the
+ *                      PK context is closed with mbedtls_pk_free(). As an
+ *                      exception, it's ok to call mbedtls_pk_free() itself
+ *                      even if the Cryptoki handles have become invalid.
+ */
+int mbedtls_pk_setup_pkcs11( mbedtls_pk_context *ctx,
+                             CK_SESSION_HANDLE hSession,
+                             CK_OBJECT_HANDLE hPublicKey,
+                             CK_OBJECT_HANDLE hPrivateKey );
+
+/**
+ * \brief               Import a transparent key into a PKCS#11 token
+ *
+ *                      This function imports a PK object containing a
+ *                      public key or a private-public key pair into a
+ *                      PKCS#11 token. 
+ *
+ * \param ctx           PK context, which must contain a transparent pk
+ *                      object (type \c MBEDTLS_PK_RSA,
+ *                      \c MBEDTLS_PK_RSASSA_PSS, \c MBEDTLS_PK_ECKEY or
+ *                      \c MBEDTLS_PK_ECDSA).
+ * \param flags         Mask of \c MBEDTLS_PKCS11_FLAG_XXX and
+ *                      \c MBEDTLS_PK_FLAG_XXX, applying as follows:
+ *                      - \c MBEDTLS_PKCS11_FLAG_TOKEN: PKCS#11 \c CKA_TOKEN
+ *                        flag: if set, import as token object; if clear,
+ *                        import as session object.
+ *                      - \c MBEDTLS_PK_FLAG_EXTRACTABLE: PKCS#11
+ *                        \c CKA_EXTRACTABLE flag: if set, the key will be
+ *                        extractable at least in wrapped form; if clear,
+ *                        the key will not be extractable at all.
+ *                      - \c MBEDTLS_PK_FLAG_SENSITIVE: PKCS#11
+ *                        \c CKA_SENSITIVE flag: if set, the key will be
+ *                        not be extractable in plain form; if clear, the
+ *                        key will be extractable at least in wrapped form.
```

Comment:
I don't see where the contradiction is. Note that the flags take permissiveness in opposite direction (EXTRACTABLE is permissive if set, SENSITIVE is permissive if unset). I'm not sure offhand that this flag is irrelevant if EXTRACTABLE is clear, I'd have to check the Cryptoki standard, which I'll do in a later documentation polishing pass.

Reply:
According to the current documentation, if both EXTRACTABLE and SENSITIVE are clear, then the key is both "not extractable at all" and "extractable at least in wrapped form", which is where I see the contradiction.

Reply:
Regarding the documentation polishing pass, I agree if it's really about polishing (like the other comment about whether "wrapped" is standard enough or needs to be explained), but when it's about stating contradictory things, like this case, I think it needs to be done before merging.

Reply:
After checking in the PKCS#11 spec:

* sensitive=false, extractable=false: wrapping forbidden, plaintext export forbidden
* sensitive=true, extractable=false: wrapping forbidden, plaintext export forbidden
* sensitive=false, extractable=true: plaintext export allowed
* sensitive=true, extractable=true: wrapping allowed, plaintext export forbidden

So the documentation should be:
```
 *                      - #MBEDTLS_PK_FLAG_EXTRACTABLE: PKCS#11
 *                        \c CKA_EXTRACTABLE flag: if set, the private key
 *                        will be extractable at least in wrapped form; if
 *                        clear, the key will not be extractable at all.
 *                      - #MBEDTLS_PK_FLAG_SENSITIVE: PKCS#11
 *                        \c CKA_SENSITIVE flag: if set, the private key
 *                        will not be extractable in plain form; if clear,
 *                        the key will be extractable in plain form if
 *                        #MBEDTLS_PK_FLAG_EXTRACTABLE is set.
```

I also corrected “key” to “private key”: the public part is always extractable.

Also references to macros defined in this file should use `#`, not `\c`, so that they're hyperlinked.

Reply:
Done